Add campaign tracking parameters for all links within Matomo app that link to matomo.org

bx80 opened the pull request:
### Description:

Fixes #10336

This PR provides a way to add campaign parameters to matomo.org links in the application which can be used to track the source of traffic to online help documents and help us improve them.

For links in **twig templates** this can done by using a new twig filter:

`{{ 'https://matomo.org/url'|trackmatomolink }}`

or with custom campaign parameters:

`{{ 'https://matomo.org/url'|trackmatomolink('MyCampaign', 'MySource', 'MyMedium') }}`
 
For links in **PHP code** the following method can be used directly:

`$url = Url::addCampaignParametersToMatomoLink('https://matomo.org/url');`

or with customer campaign parameters:

`$url = Url::addCampaignParametersToMatomoLink('https://matomo.org/url', 'MyCampaign', 'MySource', 'MyMedium');`

For links in **vue.js components**:

Change URLs from:

```
return translate(
    'Goals_LearnMoreAboutGoalTrackingDocumentation', 
    '<a target="_blank" rel="noreferrer noopener" ' 
       + 'href="https://matomo.org/docs/tracking-goals-web-analytics/">', 
    '</a>', ); 
```
to
```
import { externalLink } from 'CoreHome';
...
return translate(
    'Goals_LearnMoreAboutGoalTrackingDocumentation', 
    externalLink('https://matomo.org/docs/tracking-goals-web-analytics/'), 
    '</a>', 
);
```
Additionally `externalRawLink()` may be used to return a raw URL without the HTML tag.

For links in **vanilla JavaScript**:

```
var link =  _pk_externalRawLink('https://matomo.org/faq/how-to/faq_54/') 
```


Functionality may be disabled globally by setting a new config option:

```
[General]
disable_tracking_matomo_app_links = 1
```

Existing URL manipulation methods in `core/Url` and `core/UrlHelper` as used so that existing parameters and other URL variations are correctly handled.

The existing `UrlHelper::getPathAndQueryFromUrl()` method has been enhanced to support adding additional parameters which makes it consistent with `UrlHelper::getQueryFromUrl()`. 

During testing I noticed that `UrlHelper::getPathAndQueryFromUrl()` was unable correctly parse URLs with both an `#anchor` and URL parameters, I've improved the method so that it can preserve the anchor but since this change of behavior would be a breaking change I've made it optional via a new `$preserveAnchor` boolean parameter which defaults to the old behavior of ignoring the `#anchor`

Tests have been updated to test the new functionality in UrlHelper, new tests been added to test `addCampaignParametersToMatomoLink()` functionality with a selection of different URLs.

Any existing campaign parameters on matomo.org link have been replaced, with the exception of the URL for professional services adverts which will still use `App_ProfessionalServices` as the campaign, source will show cloud/on-premise and medium will be as before.

The following matomo.org links have *not* been updated to add campaign parameters:
- Non-clickable URLs shown in source code
- Downloads links for builds.matomo.org
- Links in the CustomVariables plugin 
- URLs shown in logs 
- URLs shown on the command line
- RSS feed links

A developer guide [here](https://github.com/matomo-org/developer-documentation/pull/764) needs review.

There are 13 dependent PRs for submodules and plugins which also need reviewing:

Plugins:
- https://github.com/innocraft/plugin-WooCommerceAnalytics/pull/25
- https://github.com/innocraft/plugin-SEOWebVitals/pull/34
- https://github.com/innocraft/plugin-SearchEngineKeywordsPerformance/pull/273
- https://github.com/innocraft/plugin-MultiChannelConversionAttribution/pull/81
- https://github.com/innocraft/plugin-CustomReports/pull/188
- https://github.com/innocraft/plugin-Cohorts/pull/65
- https://github.com/innocraft/plugin-AdvertisingConversionExport/pull/68
- https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/pull/465

Sub-modules:
- https://github.com/matomo-org/tag-manager/pull/701 
- https://github.com/matomo-org/plugin-MarketingCampaignsReporting/pull/137
- https://github.com/matomo-org/plugin-Provider/pull/50
- https://github.com/matomo-org/plugin-SecurityInfo/pull/86
- https://github.com/matomo-org/plugin-VisitorGenerator/pull/132

:warning: Before merging this PR sub-module references will need to be reset to the 5.x-dev branch :warning: 

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
Add matomo.org link campaign tracking functionality and tests

bx80 commented:
@sgiehl @michalkleiner @mneudert Do you have any thoughts on the approach here so far? 

The next step will be to update a couple of thousand matomo.org URLs in the main application and plugins to use the filter/method, so I want to be extra sure that bit only needs to be done the once! :slightly_smiling_face:

michalkleiner commented:
Nice, I like the approach of using a filter, that makes it manageable from a single place. However, saying that, I think we should keep the option to provide the module and action manually from some places and also allow sending the campaign params even when no module or action is presented, it might still be valuable to know a link comes from an instance.

So I'd suggest allowing `addCampaignParametersToMatomoLink` to also support manual overrides for the campaign via parameters for source and medium values if needed. Otherwise the overall approach looks ok to me. 

Should this also be behind a config param if someone wants to stay under the radar with on-premise installs?

bx80 commented:
> I think we should keep the option to provide the module and action manually from some places and also allow sending the campaign params even when no module or action is presented, it might still be valuable to know a link comes from an instance.

Good idea, this could also be useful for more fine-grained analysis if links appears repeatedly in the same module and action. I'll add support to optionally override the campaign parameters for `addCampaignParametersToMatomoLink` and the twig filter.

> Should this also be behind a config param if someone wants to stay under the radar with on-premise installs?

Sounds like a sensible option to add, perhaps helpful for organisations that have security policies preventing any external disclosure of deployed software? (though lots of random direct requests to Matomo help urls might be a give away!)

Commit:
Added optional override parameters for trackmatomolink, added new config option to disable matomo.org link campaign tracking, updated tests

bx80 commented:
I've added support for optional campaign parameter overrides, a global config option to disable functionality and additional tests to cover these changes. Since there is no other feedback on the general approach then I will now start converting all existing in app matomo.org links.

bx80 commented:
We have a lot of places in twig templates where the URL and link html are being passed together to the twig `translate` filter like this: 

`{{ 'API_TString'|translate("<a target='_blank' rel='noreferrer' href='https://matomo.org/link'>","</a>")|raw }}`

Since filters can't be used as parameters in other filters and the `trackmatomolink` filter expects a clean URL without HTML link markup, I've added an new twig function called `externallink` which will accept a raw URL as a parameter and return it wrapped in the normal `<a target='_blank' rel='noreferrer' ... >` link markup which is used for all external links. This function can then apply the tracking parameters if the link is for a matomo.org domain.

So the above twig will be rewritten to:

`{{ 'API_TString'|translate(externallink("https://matomo.org/link"),"</a>")|raw }}`

which is cleaner and has the advantage of centralizing any external link markup used in twig templates.

bx80 commented:
We also have [places](https://github.com/matomo-org/matomo/blob/8770cde10eafa3f9c22e3030b9ac33a26a175eff/plugins/CoreHome/templates/_donate.twig#L8) where additional html tags are being passed to the translate filter in the same parameter as the link tag:

`{{ 'API_TString'|translate("<a target='_blank' rel='noreferrer' href='https://matomo.org/link'><strong>","</strong></a>")|raw }}`

I've added an optional extra parameter to externallink to support this, it's not ideal but avoids having to change existing translation strings:

`{{ 'API_TString'|translate(externallink("https://matomo.org/link", "<strong>"),"</strong></a>")|raw }}`




 <p>{{ 'CoreHome_DonateCall3'|translate('<strong>','</strong>', externallink('https://matomo.org/recommends/premium-plugins/', '<strong>'), '</strong></a>')|raw }}</p>

michalkleiner commented:
Could we use [string interpolation](https://twig.symfony.com/doc/1.x/templates.html#string-interpolation) or the [tilde operator](https://twig.symfony.com/doc/3.x/templates.html#other-operators) to concat strings without the need to add them as extra params to the method?

Commit:
Added twig functions to format external urls, added missing matomo.org subdomains to allow list

Commit:
Revert unnecessary extra parameters for externallink twig function in favour of using twig string concatenation

bx80 commented:
> Could we use [string interpolation](https://twig.symfony.com/doc/1.x/templates.html#string-interpolation) or the [tilde operator](https://twig.symfony.com/doc/3.x/templates.html#other-operators) to concat strings without the need to add them as extra params to the method?

There are only a few of these, so that sounds like a nicer approach, I've removed the extra parameters :+1:

Commit:
Update all matomo.org links in twig templates

Commit:
Add an externalLink() function for vue.js to easily add campaign parameters to matomo.org urls in the front end

Commit:
Update vue.js externalLink() function to support the disable_tracking_matomo_app_links config option

Commit:
Attempt to keep linter happy

Commit:
More linter tweaks

bx80 commented:
@michalkleiner @sgiehl Now that the twig templates have been updated I've moved on to links in vue.js components. 

The backend `Url::addCampaignParametersToMatomoLink()` method that was added for twig/php can't be used by vue without making an API call before rendering each link which isn't really an option. So I've added a new vue.js global function called `externalLink(url)` which will accept a raw URL, apply the campaign parameters if necessary and return a formatted HTML link tag.

This seems to be working well and can be used in a vue component by adding:

`import { externalLink } from 'CoreHome';`

and changing urls from:

`return translate('Goals_LearnMoreAboutGoalTrackingDocumentation',
        '<a target="_blank" rel="noreferrer noopener" '
        + 'href="https://matomo.org/docs/tracking-goals-web-analytics/">',
        '</a>',
      );
`
to 
`
return translate(
        'Goals_LearnMoreAboutGoalTrackingDocumentation',
        externalLink('https://matomo.org/docs/tracking-goals-web-analytics/'),
        '</a>',
      );
`

The function also includes parameter overrides if we need to explicitly set the campaign parameters in a certain place.

I'm looking for some feedback on this approach generally, and specifically:
1) We need to detect the cloud plugin in order to set the campaign source parameter value, I've done this by checking for the existence of `window.Cloud` which will be the Cloud object if the plugin umd is loaded, this mirrors the backend approach of checking for the activated plugin. Is there a better/more reliable way to detect the cloud plugin loaded state in the front end?
2) We now have a `disable_tracking_matomo_app_links` config setting which can be used to disable the use of matomo URL campaign parameters. To make this config setting available at the front end without requiring another API call on every page load, I've added it as a property of the `piwik` object populated via the `_jsGlobalVariables.twig` template. This is consistent with how we make other config options globally available, but once again, is there a better way?

bx80 committed:
built vue files

Commit:
Bump stalled tests

sgiehl commented on plugins/CoreHome/vue/src/externalLink.ts:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  // Check if matomo.org domain
+  const domains = ['matomo.org', 'www.matomo.org', 'developer.matomo.org', 'plugins.matomo.org'];
```
Checking for `matomo.org` actually already includes all following checks. So no need for checking them as well...

bx80 replied:
Good point :+1: This array was copied from a similar function in `addCampaignParametersToMatomoLink` which does the check differently, I'll reduce it to just `matomo.org` but keep it as an array so additional domains can easily be added in the future.

sgiehl commented on plugins/CoreHome/vue/src/externalLink.ts:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  /* eslint-disable prefer-template */
+  return '<a target="_blank" rel="noreferrer noopener" href="' + returnUrl + '">';
```
I don't think we should include the html tag here. There might be cases were you either need additional attributes on the link or where such a url should be used to open a new tab directly in js or similar.

bx80 replied:
There are two TypeScript methods the same as for the twig filters, `externalLink` (which returns HTML) and `externalRawLink` (which just returns the raw url). Having the html tag for external links generated in one place avoids duplication and could make it easier to change in the future.

sgiehl commented on plugins/CoreVue/types/index.d.ts:
```diff
@@ -257,6 +258,7 @@ declare global {
     Piwik_Transitions: TransitionsGlobal;
     SegmentedVisitorLog: SegmentedVisitorLogService;
     DataTable_RowActions_Registry: DataTableRowActionsRegisteryService;
+    Cloud
```
doesn't this lack a type definition?

bx80 replied:
:+1: I've now set this to `any`, it'll still need testing with the cloud plugin to make sure having a property definition doesn't interfere with loading the umd.

sgiehl commented on plugins/Feedback/templates/index.twig:
```diff
@@ -56,50 +53,50 @@
-                    "<a target='_blank' rel='noreferrer noopener' href='https://matomo.org/get-involved/'>",
+                    externallink("https://matomo.org/get-involved/"),
                     "</a>",
                     "<a href='" ~ linkTo({'module':'CoreAdminHome', 'action': 'home'})~"'>",
                     "</a>",
-                    "<a target='_blank' rel='noreferrer noopener' href='https://matomo.org/team/'>",
+                    externallink("https://matomo.org/team/"),
```
this is a twig template, don't think using the javascript method would work correctly here.

bx80 replied:
There are both twig template filters `externallink` and `externalrawlink`, and also TypeScript functions `externalLink` and `externalRawLink` added in this PR (twig stuff was in the first commit). So here it's the twig filters that are being called inside `{{` braces `}}`

sgiehl reviewed (commented):
Had a very rough look through the changes to validate the approach. Looks like you are using the javascript method quite often inside twig templates, which logically doesn't make sense. But otherwise the approach looks fine.

Commit:
Remove unnecessary domains from externalLink list, add missing type definition, update and fix privacy setting link

mneudert commented on plugins/CoreHome/vue/src/externalLink.ts:
```diff
@@ -0,0 +1,106 @@
+    return '';
+  }
+
+  // Check if matomo.org domain
```
Could we simplify the code to work on an URL object?

```javascript
const returnURL = new URL(url);
const validDomain = returnURL.host.endsWith('matomo.org');

if (validDomain && module && ...) {
  ...
  returnUrl.searchParams.set('mtm_campaign', campaign);
  ...
}

return returnUrl.toString();
```

That should achieve the same result but we could remove the Regex work done to set the parameters.

bx80 replied:
Thanks @mneudert, that's a much cleaner approach than the custom regex function :+1:

Commit:
Improve externalLink TypeScript method to use URL object for domain manipulation instead of custom regex

Commit:
Make externalLink and externalRawLink TypeScript functions variadic

bx80 committed:
built vue files

Commit:
Update matomo.org links in vue components with campaign parameters

Commit:
Merge remote-tracking branch 'origin/m10336-ui-links-add-campaign' into m10336-ui-links-add-campaign

# Conflicts:
#	plugins/CoreHome/vue/dist/CoreHome.umd.js
#	plugins/CoreHome/vue/dist/CoreHome.umd.min.js
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.js
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.min.js

Commit:
Move externalRawLink function to piwik helper to make it available for vanilla JS, update matomo.org links in JS

Commit:
Update matomo.org links in PHP classes to add campaign parameters

sgiehl commented on core/Url.php:
```diff
@@ -829,4 +829,59 @@ public static function getHostFromServerNameVar()
+     *
+     * @return string|null      www.matomo.org/faq/123?mtm_campaign=Matomo_App&mtm_source=Matomo_App_OnPremise&mtm_medium=CoreAdminHome.trackingCodeGenerator
+     */
+    public static function addCampaignParametersToMatomoLink(?string $url = null, ?string $campaign = null,
```
As you seem to use that method in all plugins as well, it will be required to update the version in `Version.php` and the minimum required version in all plugins that use this new method.

Commit:
Increment version

Commit:
Update submodule references

Commit:
Fix for infinite loop when processing URLs in config settings

Commit:
Update mtm_medium to include 'App.' prefix, linter pacification changes for externalLink

Commit:
Fix link formatting in GeoIp2 vue component, update system tests

bx80 committed:
built vue files

Commit:
Update advertising test, fix missing import in Tour

Commit:
Merge remote-tracking branch 'origin/m10336-ui-links-add-campaign' into m10336-ui-links-add-campaign

Commit:
UI test fixes for tracking code twig templates and license key formatting

bx80 committed:
built vue files

Commit:
Bump stalled tests

Commit:
Merge remote-tracking branch 'origin/m10336-ui-links-add-campaign' into m10336-ui-links-add-campaign

michalkleiner commented on plugins/Feedback/vue/src/RateFeature/RateFeature.vue:
```diff
@@ -148,8 +148,7 @@
           <p v-if="likeReason || dislikeReason"
-            v-html="$sanitize(translate('Feedback_Policy',`
-            <a rel='nofollow' href='https://matomo.org/privacy-policy/' target='_blank'>`,'</a>'))">
+            v-html="$sanitize(translate('Feedback_Policy', externalLink('https://matomo.org/privacy-policy/'),'</a>'))">
```
```suggestion
            v-html="$sanitize(translate('Feedback_Policy', externalLink('https://matomo.org/privacy-policy/'), '</a>'))">
```

michalkleiner commented on core/API/Request.php:
```diff
@@ -474,7 +474,8 @@ public static function checkTokenAuthIsNotLimited($module, $action)
         if (Access::getInstance()->hasSuperUserAccess()) {
-            $ex = new \Piwik\Exception\Exception(Piwik::translate('Widgetize_TooHighAccessLevel', ['<a href="https://matomo.org/faq/troubleshooting/faq_147/" rel="noreferrer noopener">', '</a>']));
+            $ex = new \Piwik\Exception\Exception(Piwik::translate('Widgetize_TooHighAccessLevel',
+                ['<a href="' . Url::addCampaignParametersToMatomoLink('https://matomo.org/faq/troubleshooting/faq_147/') . '" rel="noreferrer noopener">', '</a>']));
```
It nearly feels like we could have an equal method returning the opening HTML element for an external link with the noreferrer noopener rel attribute even on the PHP side.

I get it would slightly obscure that it actually produces an HTML tag, but would be consistent with how it'll be done in Vue and shorten the lines a bit while avoiding string concatenations in tens of files with potential for human errors.

michalkleiner replied:
Saying that, we can probably iterate on that, it doesn't need to hold up the PR.

michalkleiner commented on core/Nonce.php:
```diff
@@ -103,7 +103,7 @@ public static function verifyNonceWithErrorMessage($id, $cnonce, $allowedReferre
             $additionalErrors =  '<br/><br/>' . Piwik::translate('Login_InvalidNonceSSLMisconfigured',
                 array(
-                  '<a target="_blank" rel="noreferrer noopener" href="https://matomo.org/faq/how-to/faq_91/">',
+                  '<a target="_blank" rel="noreferrer noopener" href="' . Url::addCampaignParametersToMatomoLink('https://matomo.org/faq/how-to/faq_91/') . '">',
```
Just a suggestion what this could look like:
```suggestion
                  Url::getExternalMatomoLinkTag('https://matomo.org/faq/how-to/faq_91/'),
```

bx80 replied:
This would be a nice code improvement, I've created a [new issue](https://github.com/matomo-org/matomo/issues/21411) for it. 
`addCampaignParametersToMatomoLink()` will ignore non-matomo domains, so we could use the same method for all external links as there are quite a external links which go to non-Matomo domains. So maybe the new method could be called `getExternalLinkTag()`.

michalkleiner commented on core/Nonce.php:
```diff
@@ -123,7 +123,7 @@ public static function verifyNonceWithErrorMessage($id, $cnonce, $allowedReferre
             if (empty($allowedReferrerHost) && !Url::isLocalUrl($referrer)) {
                 return Piwik::translate('Login_InvalidNonceReferrer', array(
-                        '<a target="_blank" rel="noreferrer noopener" href="https://matomo.org/faq/how-to-install/faq_98">',
+                        '<a target="_blank" rel="noreferrer noopener" href="' . Url::addCampaignParametersToMatomoLink('https://matomo.org/faq/how-to-install/faq_98') . '">',
                         '</a>'
```
and then we could also do an optional param which would make it return an array of the opening and the closing tags, usable with the destructuring `...` in arrays.
```suggestion
                        ...Url::getExternalMatomoLinkTag('https://matomo.org/faq/how-to-install/faq_98', true),
```

michalkleiner replied:
Might be too much magic for some people's appetite though 🤷

michalkleiner commented on plugins/CoreAdminHome/vue/src/JsTrackingCodeGenerator/JsTrackingCodeGenerator.vue:
```diff
@@ -22,19 +22,19 @@
         <br><br/>
         {{ translate('SitesManager_InstallationGuides') }} :
-        <a href="https://matomo.org/faq/new-to-piwik/how-do-i-install-the-matomo-tracking-code-on-wordpress/"
+        <a href="{{ externalRawLink('https://matomo.org/faq/new-to-piwik/how-do-i-install-the-matomo-tracking-code-on-wordpress/', 'abc123', 'def456', 'ghi789') }}"
```
```suggestion
        <a href="{{ externalRawLink('https://matomo.org/faq/new-to-piwik/how-do-i-install-the-matomo-tracking-code-on-wordpress/') }}"
```

bx80 replied:
Those test parameters shouldn't be there at all, Ive removed them :+1:

michalkleiner commented on plugins/SitesManager/SiteContentDetection/Complianz.php:
```diff
@@ -19,7 +19,7 @@ public static function getName(): string
     public static function getInstructionUrl(): ?string
     {
-        return 'https://matomo.org/faq/how-to/using-complianz-for-wordpress-consent-manager-with-matomo';
+        return Url::addCampaignParametersToMatomoLink('https://matomo.org/faq/how-to/using-complianz-for-wordpress-consent-manager-with-matomo');
```
Missing use statement

michalkleiner commented on plugins/SitesManager/SiteContentDetection/TarteAuCitron.php:
```diff
@@ -22,7 +22,7 @@ public static function getName(): string
     public static function getInstructionUrl(): ?string
     {
-        return 'https://matomo.org/faq/how-to/using-tarte-au-citron-consent-manager-with-matomo';
+        return Url::addCampaignParametersToMatomoLink('https://matomo.org/faq/how-to/using-tarte-au-citron-consent-manager-with-matomo');
```
Missing use statement

michalkleiner commented:
Went over all the link changes and majority of the functional changes, still have a few files to review but generally this looks very good!

Left a comment for discussion/further enhancement about the PHP creation of external links — for a separate ticket if decided to work on.

Commit:
Fix missing use statements, remove incorrect custom params from link

bx80 committed:
built vue files

Commit:
Change update version back to 5.0.0-rc5

michalkleiner commented on plugins/CoreHome/vue/src/externalLink.ts:
```diff
@@ -0,0 +1,47 @@
+  ...values: (string|null)[]
+): string {
+  const pkArgs = values as (string|null)[];
+  return window._pk_externalRawLink(url, pkArgs); // eslint-disable-line
```
Do we need to check if the `_pk_externalRawLink` method is available on the `window` object and if not, just return the original `url`?

bx80 replied:
I've added a check for this :+1:

michalkleiner commented on plugins/Morpheus/javascripts/piwikHelper.js:
```diff
@@ -23,6 +23,42 @@ function _pk_translate(translationStringId, values) {
+  const mediumOverride = (typeof values != 'undefined' && values.length > 2 && values[2] ? values[2] : null);
+
+  const returnURL = new URL(url);
+  const validDomain = returnURL.host.endsWith('matomo.org');
```
to be on the safe side with subdomains and not just catch `anyrandomdomainformatomo.org`
```suggestion
  const validDomain = returnURL.host === 'matomo.org' || returnURL.host.endsWith('.matomo.org');
```

michalkleiner commented on plugins/Morpheus/javascripts/piwikHelper.js:
```diff
@@ -23,6 +23,42 @@ function _pk_translate(translationStringId, values) {
+  if (validDomain && !window.piwik.disableTrackingMatomoAppLinks
+    && ((module && action) || mediumOverride)) {
+    const campaign = (campaignOverride === null ? 'Matomo_App' : campaignOverride);
+    let source = (window.Cloud === undefined ? 'OnPremise' : 'Cloud');
```
The PHP version has these as 'Matomo_App_OnPremise' and 'Matomo_App_Cloud', should they be the same here or adjust in PHP so it's consistent?

bx80 replied:
Good spot, I've updated the JS version to be consistent with PHP :+1:

michalkleiner commented on plugins/MultiSites/vue/src/Dashboard/Dashboard.vue:
```diff
@@ -8,7 +8,7 @@
     <h2 class="card-title">
       <EnrichedHeadline
-        help-url="https://matomo.org/faq/new-to-piwik/all-websites-dashboard/"
+        help-url="{{ externalLink('https://matomo.org/faq/new-to-piwik/all-websites-dashboard/') }}"
```
```suggestion
        help-url="{{ externalRawLink('https://matomo.org/faq/new-to-piwik/all-websites-dashboard/') }}"
```

michalkleiner commented on plugins/MultiSites/vue/src/Dashboard/Dashboard.vue:
```diff
@@ -153,13 +153,13 @@
                 rel="noreferrer noopener"
                 target="_blank"
-                href="https://matomo.org/faq/troubleshooting/faq_19489/"
+                href="{{ externalLink('https://matomo.org/faq/troubleshooting/faq_19489/') }}"
```
```suggestion
                href="{{ externalRawLink('https://matomo.org/faq/troubleshooting/faq_19489/') }}"
```

michalkleiner commented on plugins/MultiSites/vue/src/Dashboard/Dashboard.vue:
```diff
@@ -153,13 +153,13 @@
                 rel="noreferrer noopener"
                 target="_blank"
-                href="https://forum.matomo.org/"
+                href="{{ externalLink('https://forum.matomo.org/') }}"
```
```suggestion
                href="{{ externalRawLink('https://forum.matomo.org/') }}"
```

michalkleiner reviewed (changes requested):
Reviewed the remaining bits and have a small number of changes to be made, after that it's good to go from my perspective.

bx80 committed:
Update plugins/MultiSites/vue/src/Dashboard/Dashboard.vue

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
Update plugins/MultiSites/vue/src/Dashboard/Dashboard.vue

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
Update plugins/MultiSites/vue/src/Dashboard/Dashboard.vue

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
Update plugins/Morpheus/javascripts/piwikHelper.js

Co-authored-by: Michal Kleiner <michal@innocraft.com>

bx80 committed:
built vue files

Commit:
Add check for _pk_externalRawLink before calling, Fix inconsistent source string

Commit:
Merge remote-tracking branch 'origin/m10336-ui-links-add-campaign' into m10336-ui-links-add-campaign

Commit:
Fix linter complaint

bx80 committed:
built vue files

Commit:
Bump stalled tests

Commit:
Update submodule references to 5.x-dev branches

Commit:
Update visitorgenerator submodule reference to 5.x-dev branche
